pulser/simulation/simulation.py

```diff
@@ -15,116 +15,214 @@
+                 config: Optional[SimConfig] = None,
+                 evaluation_times: Union[float, str, ArrayLike] = 'Full'
+                 ) -> None:
+        """Instantiates a Simulation object."""
+        self.seq = sequence
         self._qdict = self._seq.qubit_info
         self._size = len(self._qdict)
         self._tot_duration = max(
             [self._seq._last(ch).tf for ch in self._seq._schedule]
         )
-
-        if not (0 < sampling_rate <= 1.0):
-            raise ValueError("The sampling rate (`sampling_rate` = "
-                             f"{sampling_rate}) must be greater than 0 and "
-                             "less than or equal to 1.")
-        if int(self._tot_duration*sampling_rate) < 4:
-            raise ValueError("`sampling_rate` is too small, less than 4 data "
-                             "points.")
         self.sampling_rate = sampling_rate
-
         self._qid_index = {qid: i for i, qid in enumerate(self._qdict)}
         self.samples: dict[str, dict[str, dict]] = {
             addr: {basis: {} for basis in ['ground-rydberg', 'digital']}
-            for addr in ['Global', 'Local']
-        }
+            for addr in ['Global', 'Local']}
         self.operators = deepcopy(self.samples)
+        self._times = self._adapt_to_sampling_rate(
+            np.arange(self._tot_duration, dtype=np.double)/1000)
+        self.evaluation_times = evaluation_times
+        self.config = config if config else SimConfig()
+        self.initial_state = 'all-ground'
+        self._collapse_ops: List[qutip.Qobj] = []
+
+    @property
+    def sampling_rate(self) -> float:
+        """Property getter for sampling_rate."""
+        return self._sampling_rate
+
+    @sampling_rate.setter
+    def sampling_rate(self, value: float) -> None:
+        if not (0 < value <= 1.0):
+            raise ValueError("`sampling_rate` must be positive and "
+                             "not larger than 1.0")
+        if int(self._tot_duration * value) < 4:
+            raise ValueError("`sampling_rate` is too small, less than 4 data "
+                             "points.")
+        self._sampling_rate: float = value
 
-        self._extract_samples()
-        self._build_basis_and_op_matrices()
-        self._construct_hamiltonian()
+    @property
+    def seq(self) -> Sequence:
+        """Property getter for sequence."""
+        return self._seq
+
+    @seq.setter
+    def seq(self, seq: Sequence) -> None:
+        if not isinstance(seq, Sequence):
+            raise TypeError("The provided sequence has to be a valid "
+                            "pulser.Sequence instance.")
+        if not seq._schedule:
+            raise ValueError("The provided sequence has no declared channels.")
+        if all(seq._schedule[x][-1].tf == 0 for x in seq._channels):
+            raise ValueError("No instructions given for the channels in the "
+                             "sequence.")
+        self._seq = seq
```

Comment:
What's the usecase for a setter for Sequence? To me, at that point it would make more sense to initialize a new Simulation

Reply:
I thought about that as well, this as a practical way to set the Sequence in init but I think it's a bad idea to leave it public.

Reply:
It's okay to leave it explicitly in the init, specially if it's not going to be used anywhere else